File: src/main/intellibitz/sted/io/FontMapXMLHandler.java
package intellibitz.sted.io;

import intellibitz.sted.fontmap.FontMap;
import intellibitz.sted.fontmap.FontMapEntry;
import intellibitz.sted.util.FileHelper;
import org.xml.sax.Attributes;
import org.xml.sax.SAXException;
import org.xml.sax.helpers.DefaultHandler;

import javax.xml.parsers.ParserConfigurationException;
import javax.xml.parsers.SAXParser;
import javax.xml.parsers.SAXParserFactory;
import java.io.IOException;
import java.io.InputStream;
import java.util.logging.Logger;

class FontMapXMLHandler
        extends DefaultHandler {
    private FontMap fontMap;
    private FontMapEntry fontMapEntry;
    private static final Logger logger =
            Logger.getLogger("intellibitz.sted.io.FontMapXMLHandler");
    private boolean valid;

    FontMapXMLHandler() {
        super();
    }

    FontMapXMLHandler(FontMap fontMap) {
        this();
        this.fontMap = fontMap;
    }

    public void read(FontMap fontMap)
            throws IOException, ParserConfigurationException, SAXException {
        logger.entering(getClass().getName(), "read", fontMap);
        this.fontMap = fontMap;
        final SAXParserFactory saxParserFactory =
                SAXParserFactory.newInstance();
        saxParserFactory.setValidating(true);
        final SAXParser saxParser = saxParserFactory.newSAXParser();
        logger.info("parsing file - " + fontMap.getFontMapFile());
        final InputStream inputStream =
                FileHelper.getInputStream(fontMap.getFontMapFile());
        if (inputStream == null) {
            logger.severe("InputStream NULL - Cannot Read File: " +
                    fontMap.getFontMapFile());
            throw new IOException("InputStream NULL - Cannot Read File: " +
                    fontMap.getFontMapFile());
        } else {
            saxParser.parse(inputStream, this);
        }
        logger.exiting(getClass().getName(), "read");
    }

    public void startElement(String uri, String localName, String qName,
                             Attributes attributes)
            throws SAXException {
        if ("fontmap".equals(qName)) {
            valid = true;
        } else if ("font_from".equals(qName)) {
            fontMap.setFont1Path(attributes.getValue("path"));
            fontMap.setFont1(attributes.getValue("value"));
        } else if ("font_to".equals(qName)) {
            fontMap.setFont2Path(attributes.getValue("path"));
            fontMap.setFont2(attributes.getValue("value"));
        } else if ("font_entry".equals(qName)) {
            fontMapEntry = new FontMapEntry();
        } else if ("entry_from".equals(qName)) {
            fontMapEntry.setFrom(attributes.getValue("value"));
        } else if ("entry_to".equals(qName)) {
            fontMapEntry.setTo(attributes.getValue("value"));
        } else if ("begins_with".equals(qName)) {
            fontMapEntry.setBeginsWith(attributes.getValue("value"));
        } else if ("ends_with".equals(qName)) {
            fontMapEntry.setEndsWith(attributes.getValue("value"));
        } else if ("preceded_by".equals(qName)) {
            fontMapEntry.setPrecededBy(attributes.getValue("value"));
        } else if ("followed_by".equals(qName)) {
            fontMapEntry.setFollowedBy(attributes.getValue("value"));
        } else if ("conditional".equals(qName)) {
            fontMapEntry.setConditional(attributes.getValue("value"));
        }
    }

    public void endElement(String uri, String localName, String qName)
            throws SAXException {
        if (!valid) {
            throw new SAXException(
                    "Invalid FontMap.. Did not have a valid Header");
        }
        if ("font_entry".equals(qName)) {
            fontMap.getEntries().add(fontMapEntry);
        }
    }


}

File: src/main/intellibitz/sted/io/FontMapXMLHandler.kt
package sted.io

import org.xml.sax.Attributes
import org.xml.sax.SAXException
import org.xml.sax.helpers.DefaultHandler
import sted.fontmap.FontMap
import sted.fontmap.FontMapEntry
import sted.io.FileHelper.getInputStream
import java.io.IOException
import java.util.logging.Logger
import javax.xml.parsers.ParserConfigurationException
import javax.xml.parsers.SAXParserFactory

internal class FontMapXMLHandler : DefaultHandler() {
    lateinit var fontMap: FontMap
    lateinit var fontMapEntry: FontMapEntry
    private var valid = false

    @Throws(IOException::class, ParserConfigurationException::class, SAXException::class)
    fun read(fontMap: FontMap) {
        logger.entering(javaClass.name, "read", fontMap)
        this.fontMap = fontMap
        val saxParserFactory = SAXParserFactory.newInstance()
        saxParserFactory.isValidating = true
        val saxParser = saxParserFactory.newSAXParser()
        logger.info("parsing file - " + fontMap.fontMapFile)
        val inputStream = getInputStream(fontMap.fontMapFile)
        saxParser.parse(inputStream, this)
        logger.exiting(javaClass.name, "read")
    }

    override fun startElement(uri: String, localName: String, qName: String, attributes: Attributes) {
        when (qName) {
            "fontmap" -> {
                valid = true
            }
            "font_from" -> {
                fontMap.font1Path = attributes.getValue("path")
                fontMap.setFont1(attributes.getValue("value"))
            }
            "font_to" -> {
                fontMap.font2Path = attributes.getValue("path")
                fontMap.setFont2(attributes.getValue("value"))
            }
            "font_entry" -> {
                fontMapEntry = FontMapEntry()
            }
            "entry_from" -> {
                fontMapEntry.from = attributes.getValue("value")
            }
            "entry_to" -> {
                fontMapEntry.to = attributes.getValue("value")
            }
            "begins_with" -> {
                fontMapEntry.setBeginsWith(attributes.getValue("value"))
            }
            "ends_with" -> {
                fontMapEntry.setEndsWith(attributes.getValue("value"))
            }
            "preceded_by" -> {
                fontMapEntry.precededBy = attributes.getValue("value")
            }
            "followed_by" -> {
                fontMapEntry.followedBy = attributes.getValue("value")
            }
            "conditional" -> {
                fontMapEntry.setConditional(attributes.getValue("value"))
            }
        }
    }

    @Throws(SAXException::class)
    override fun endElement(uri: String, localName: String, qName: String) {
        if (!valid) {
            throw SAXException(
                "Invalid FontMap.. Did not have a valid Header"
            )
        }
        if ("font_entry" == qName) {
            fontMap.entries.add(fontMapEntry)
        }
    }

    companion object {
        private val logger = Logger.getLogger("sted.io.FontMapXMLHandler")
    }
}